File: stormbird/src/lifting_line/simulation_builder.rs
// Copyright (C) 2024, NTNU
// Author: Jarle Vinje Kramer <jarlekramer@gmail.com; jarle.a.kramer@ntnu.no>
// License: GPL v3.0 (see separate file LICENSE or https://www.gnu.org/licenses/gpl-3.0.html)

//! A builder for the [Simulation] struct.

use serde::{Serialize, Deserialize};

use crate::lifting_line::prelude::*;
use crate::lifting_line::wake::{
    frozen_wake::FrozenWake,
    dynamic_wake::builder::DynamicWakeBuilder,
};

use crate::lifting_line::wake::settings::QuasiSteadyWakeSettings;

use super::simulation::Simulation;

use crate::error::Error;

#[derive(Debug, Serialize, Deserialize, Default, Clone)]
#[serde(deny_unknown_fields)]
/// Settings for a quasi-steady simulation.
pub struct QuasiSteadySettings {
    #[serde(default)]
    pub solver: QuasiSteadySolverBuilder,
    #[serde(default)]
    pub wake: QuasiSteadyWakeSettings,
}

#[derive(Debug, Serialize, Deserialize, Default, Clone)]
#[serde(deny_unknown_fields)]
/// Settings for a dynamic simulation.
pub struct DynamicSettings {
    #[serde(default)]
    pub solver: Solver,
    #[serde(default)]
    pub wake: DynamicWakeBuilder,
}


#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
/// Enum that determines what type of simulation to run. 
/// 
/// Both quasi-steady and dynamic simulations are supported. The settings for each simulation type 
/// is stored as a member variable of each variant.
pub enum SimulationSettings {
    QuasiSteady(QuasiSteadySettings),
    Dynamic(DynamicSettings),
}

impl Default for SimulationSettings {
    fn default() -> Self {
        SimulationSettings::QuasiSteady(QuasiSteadySettings::default())
    }
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(deny_unknown_fields)]
/// Builder for the [Simulation] struct.
pub struct SimulationBuilder {
    pub line_force_model: LineForceModelBuilder,
    #[serde(default)]
    pub simulation_settings: SimulationSettings,
}

impl SimulationBuilder {
    pub fn new(
        line_force_model: LineForceModelBuilder,
        simulation_settings: SimulationSettings,
    ) -> Self {
        SimulationBuilder {
            line_force_model,
            simulation_settings,
        }
    }

    /// Creates a new simulation builder by parsing the the string as a JSON object. The parsing is
    /// done using the serde_json library.
    pub fn new_from_string(setup_string: &str) -> Result<Self, Error> {
        let builder = serde_json::from_str(setup_string)?;
        
        Ok(builder)
    }

    /// Creates a new simulation builder by reading the file at the given path and parsing the 
    /// content as a JSON object. The parsing is done using the [SimulationBuilder::new_from_string]
    /// method.
    pub fn new_from_file(file_path: &str) -> Result<Self, Error> {
        let string = std::fs::read_to_string(file_path)?;

        Self::new_from_string(&string)
    }

    /// Builds the [Simulation] struct based on the current state of the builder.
    pub fn build(&self) -> Simulation {
        let line_force_model = self.line_force_model.build();

        let nr_of_lines = line_force_model.nr_span_lines();

        let wake_data = match &self.simulation_settings {
            SimulationSettings::Dynamic(settings) => {
                WakeData::Dynamic(
                    settings.wake.build(
                        &line_force_model
                    )
                )
            },
            SimulationSettings::QuasiSteady(settings) => {
                WakeData::QuasiSteady(
                    settings.wake.clone()
                )
            }
        };

        let frozen_wake = FrozenWake::initialize(nr_of_lines);

        let solver = match &self.simulation_settings {
            SimulationSettings::Dynamic(settings) => {
                settings.solver.clone()
            },
            SimulationSettings::QuasiSteady(settings) => {
                settings.solver.build()
            }
        };

        let previous_circulation_strength = vec![0.0; nr_of_lines];

        let flow_derivatives = FlowDerivatives::new(
            &vec![SpatialVector::from([0.0, 0.0, 0.0]); nr_of_lines]
        );

        Simulation {
            line_force_model,
            flow_derivatives,
            wake_data,
            frozen_wake,
            solver,
            previous_circulation_strength,
            first_time_step_completed: false,
        }
    }
}

File: interfaces/stormbird_setup/src/stormbird_setup/direct_setup/lifting_line/simulation_builder.py
'''
Copyright (C) 2024, NTNU
Author: Jarle Vinje Kramer <jarlekramer@gmail.com; jarle.a.kramer@ntnu.no>
License: GPL v3.0 (see separate file LICENSE or https://www.gnu.org/licenses/gpl-3.0.html)
'''
from typing import Any

from pydantic import model_serializer, model_validator

from ...base_model import StormbirdSetupBaseModel
from ..line_force_model import LineForceModelBuilder

from .solver import Linearized, SimpleIterative
from .wake import QuasiSteadyWakeSettings, DynamicWakeBuilder

class QuasiSteadySettings(StormbirdSetupBaseModel):
    solver: Linearized | SimpleIterative = Linearized()
    wake: QuasiSteadyWakeSettings  = QuasiSteadyWakeSettings ()
    
    @model_validator(mode='before')
    @classmethod
    def deserialize_solver(cls, data: Any) -> Any:
        if isinstance(data, dict) and 'solver' in data:
            solver_data = data['solver']
            if isinstance(solver_data, dict):
                if 'Linearized' in solver_data:
                    data['solver'] = Linearized(**solver_data['Linearized'])
                elif 'SimpleIterative' in solver_data:
                    data['solver'] = SimpleIterative(**solver_data['SimpleIterative'])
        return data

    @model_serializer
    def ser_model(self):
        solver_dict = self.solver.model_dump()
        wake_dict = self.wake.model_dump()

        if isinstance(self.solver, Linearized):
            return {
                "solver": {
                    "Linearized": solver_dict
                },
                "wake": wake_dict
            }
        elif isinstance(self.solver, SimpleIterative):
            return {
                "solver": {
                    "SimpleIterative": solver_dict
                },
                "wake": wake_dict
            }

class DynamicSettings(StormbirdSetupBaseModel):
    solver: SimpleIterative | Linearized = SimpleIterative()
    wake: DynamicWakeBuilder = DynamicWakeBuilder()
    
    @model_validator(mode='before')
    @classmethod
    def deserialize_solver(cls, data: Any) -> Any:
        if isinstance(data, dict) and 'solver' in data:
            solver_data = data['solver']
            if isinstance(solver_data, dict):
                if 'Linearized' in solver_data:
                    data['solver'] = Linearized(**solver_data['Linearized'])
                elif 'SimpleIterative' in solver_data:
                    data['solver'] = SimpleIterative(**solver_data['SimpleIterative'])
        return data

    @model_serializer
    def ser_model(self):
        solver_dict = self.solver.model_dump()
        wake_dict = self.wake.model_dump()

        if isinstance(self.solver, Linearized):
            return {
                "solver": {
                    "Linearized": solver_dict
                },
                "wake": wake_dict
            }
        elif isinstance(self.solver, SimpleIterative):
            return {
                "solver": {
                    "SimpleIterative": solver_dict
                },
                "wake": wake_dict
            }

class SimulationBuilder(StormbirdSetupBaseModel):
    line_force_model: LineForceModelBuilder = LineForceModelBuilder()
    simulation_settings: QuasiSteadySettings | DynamicSettings = QuasiSteadySettings()
    
    @model_validator(mode='before')
    @classmethod
    def deserialize_simulation_settings(cls, data: Any) -> Any:
        if isinstance(data, dict) and 'simulation_settings' in data:
            settings_data = data['simulation_settings']
            if isinstance(settings_data, dict):
                if 'QuasiSteady' in settings_data:
                    data['simulation_settings'] = QuasiSteadySettings(**settings_data['QuasiSteady'])
                elif 'Dynamic' in settings_data:
                    data['simulation_settings'] = DynamicSettings(**settings_data['Dynamic'])
        return data

    @model_serializer
    def ser_model(self):
        line_force_model_dict = self.line_force_model.model_dump()
        simulation_settings_dict = self.simulation_settings.model_dump()

        if isinstance(self.simulation_settings, QuasiSteadySettings):
            return {
                "line_force_model": line_force_model_dict,
                "simulation_settings": {
                    "QuasiSteady": simulation_settings_dict
                }
            }
        elif isinstance(self.simulation_settings, DynamicSettings):
            return {
                "line_force_model": line_force_model_dict,
                "simulation_settings": {
                    "Dynamic": simulation_settings_dict
                }
            }
